File: 10_ransom_note.c
#include <stdio.h>
#include <string.h>

int canConstruct(char* ransomNote, char* magazine) {
    int freq[256] = {0};
    for (int i = 0; magazine[i]; i++)
        freq[(unsigned char)magazine[i]]++;
    for (int i = 0; ransomNote[i]; i++) {
        if (--freq[(unsigned char)ransomNote[i]] < 0)
            return 0;
    }
    return 1;
}

int main() {
    char ransomNote[] = "aa";
    char magazine[] = "aab";
    printf("%d\n", canConstruct(ransomNote, magazine));
    return 0;
}

File: 10_ransom_note.py
from collections import Counter

def can_construct_ransom(ransomNote, magazine):
    ransom_count = Counter(ransomNote)
    magazine_count = Counter(magazine)
    for c in ransom_count:
        if ransom_count[c] > magazine_count.get(c, 0):
            return False
    return True

if __name__ == "__main__":
    ransomNote = "aa"
    magazine = "aab"
    print(can_construct_ransom(ransomNote, magazine))
